### PR TITLE
multiple database support for UCRs

### PR DESCRIPTION
**Pull request by czue**
properly use report-defined engines in pillow and reports, and add tests for these things

probably easiest read commit-by-commit, though might want to read the new tests as a standalone file

@snopoke @NoahCarnahan

### CODE REVIEW

File: corehq/apps/userreports/tests/test_multi_db.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
+import uuid
+from django.conf import settings
+from django.test import TestCase
+from mock import patch
+from sqlalchemy import create_engine
+from sqlalchemy.exc import ProgrammingError
+from corehq.apps.userreports.models import DataSourceConfiguration, ReportConfiguration
+from corehq.apps.userreports.pillow import ConfigurableIndicatorPillow
+from corehq.apps.userreports.reports.data_source import ConfigurableReportDataSource
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_multi_db.py]: F401 'ConfigurableReportDataSource' imported but unused

File: corehq/apps/userreports/tests/test_multi_db.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            patch('corehq.apps.userreports.sql.adapter.get_engine_id'),
+            patch('corehq.apps.userreports.reports.data_source.get_engine_id'),
+        )
+        cls.connection_string_patch = patch('corehq.db.connection_manager.get_connection_string')
+        for engine_id_patch in cls.engine_id_patches:
+            mock_engine_id_method = engine_id_patch.start()
+            mock_engine_id_method.side_effect = lambda x: x.engine_id
+
+
+        def connection_string_for_engine(engine_id):
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_multi_db.py]: E303 too many blank lines (2)

File: corehq/apps/userreports/tests/test_multi_db.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        conn.execute('commit')
+        try:
+            conn.execute('CREATE DATABASE {}'.format(cls.db2_name))
+        except ProgrammingError:
+            # optimistically assume it failed because was already created.
+            pass
+        conn.close()
+
+        # initialize the UCR tables
+        cls.ds1_adapter= IndicatorSqlAdapter(cls.ds_1)
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_multi_db.py]: E225 missing whitespace around operator

File: corehq/apps/userreports/pillow.py [hunk 4]
```diff
@@ -45,19 +38,23 @@ def bootstrap(self, configs=None):
...
-            diffs = compare_metadata(migration_context, metadata)
-
-        tables_to_rebuild = get_tables_to_rebuild(diffs, table_map.keys())
-        for table_name in tables_to_rebuild:
-            table = table_map[table_name]
-            try:
-                self.rebuild_table(table)
-            except TableRebuildError, e:
-                notify_error(unicode(e))
+        tables_by_engine = defaultdict(lambda: {})
```
Comment (snopoke) [inline corehq/apps/userreports/pillow.py]: could be `defaultdict(dict)`

File: corehq/apps/userreports/tests/test_multi_db.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        cls.ds_2 = DataSourceConfiguration.wrap(data_source_template.to_json())
+        cls.ds_2.engine_id = 'engine-2'
+        cls.ds_2.save()
+
+        # use db1 engine to create db2 http://stackoverflow.com/a/8977109/8207
+        cls.root_engine = create_engine(settings.SQL_REPORTING_DATABASE_URL)
+        conn = cls.root_engine.connect()
+        conn.execute('commit')
+        try:
+            conn.execute('CREATE DATABASE {}'.format(cls.db2_name))
```
Comment (snopoke) [inline corehq/apps/userreports/tests/test_multi_db.py]: classic :boot: starping

File: corehq/apps/userreports/tests/test_multi_db.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        # delete data sources
+        cls.ds_1.delete()
+        cls.ds_2.delete()
+
+        # dispose secondary engine
+        cls.ds2_adapter.session_helper.engine.dispose()
+
+        # drop the secondary database
+        conn = cls.root_engine.connect()
+        conn.execute('commit')
```
Comment (snopoke) [inline corehq/apps/userreports/tests/test_multi_db.py]: Should this not be a rollback?

File: corehq/apps/userreports/tests/test_multi_db.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            conn.close()
+            cls.root_engine.dispose()
+
+    def tearDown(self):
+        self.ds1_adapter.session_helper.Session.remove()
+        self.ds2_adapter.session_helper.Session.remove()
+        self.ds1_adapter.drop_table()
+        self.ds2_adapter.drop_table()
+
+    def test_patches(self):
```
Comment (snopoke) [inline corehq/apps/userreports/tests/test_multi_db.py]: test that my tests are working :)

**Comment by snopoke** [comment]
Nothing blocking and tests passing so going to merge.

**Comment by czue** [comment]
thanks, responded to feedback here: https://github.com/dimagi/commcare-hq/pull/7537

**Pull request merged**